File: lib/oli/content/writers/html.ex
defmodule Oli.Content.Writers.HTML do
  alias Oli.Content.Writers.Context
  alias Phoenix.HTML

  @behaviour Oli.Content.Writers.Writer

  def p(%Context{} = _context, next, _) do
    ["<p>", next.(), "</p>\n"]
  end

  def h1(%Context{} = _context, next, _) do
    ["<h1>", next.(), "</h1>\n"]
  end

  def h2(%Context{} = _context, next, _) do
    ["<h2>", next.(), "</h2>\n"]
  end

  def h3(%Context{} = _context, next, _) do
    ["<h3>", next.(), "</h3>\n"]
  end

  def h4(%Context{} = _context, next, _) do
    ["<h4>", next.(), "</h4>\n"]
  end

  def h5(%Context{} = _context, next, _) do
    ["<h5>", next.(), "</h5>\n"]
  end

  def h6(%Context{} = _context, next, _) do
    ["<h6>", next.(), "</h6>\n"]
  end

  def img(%Context{} = _context, _, %{"src" => src} = attrs) do

    height_width = case attrs do
      %{"height" => height, "width" => width} -> ["height=#{height} width=#{width}"]
      _ -> []
    end

    ["<img "]
    ++ height_width
    ++ [" style=\"display: block; max-height: 500px; margin-left: auto; margin-right: auto;\" src=\"", src, "\"/>\n"]
  end

  def youtube(%Context{} = _context, _, %{"src" => src}) do
    ["""
    <iframe
      id="#{src}"
      width="640"
      height="476"
      src="https://www.youtube.com/embed/#{src}"
      frameBorder="0"
      style="display: block; margin-left: auto; margin-right: auto;"
    ></iframe>
    """]
  end

  def audio(%Context{} = _context, _, %{"src" => src}) do
    ["<audio src=\"", src, "\"/>\n"]
  end

  def table(%Context{} = _context, next, _) do
    ["<table>", next.(), "</table>\n"]
  end

  def tr(%Context{} = _context, next, _) do
    ["<tr>", next.(), "</tr>\n"]
  end

  def th(%Context{} = _context, next, _) do
    ["<th>", next.(), "</th>\n"]
  end

  def td(%Context{} = _context, next, _) do
    ["<td>", next.(), "</td>\n"]
  end

  def ol(%Context{} = _context, next, _) do
    ["<ol>", next.(), "</ol>\n"]
  end

  def ul(%Context{} = _context, next, _) do
    ["<ul>", next.(), "</ul>\n"]
  end

  def li(%Context{} = _context, next, _) do
    ["<li>", next.(), "</li>\n"]
  end

  def math(%Context{} = _context, next, _) do
    ["<div>\\[", next.(), "\\]</div>\n"]
  end

  def math_line(%Context{} = _context, next, _) do
    [next.(), "\n"]
  end

  def code(%Context{} = _context, next, %{
    "language" => _language,
    "startingLineNumber" => _startingLineNumber,
    "showNumbers" => _showNumbers,
    "caption" => _caption
  }) do
    ["<pre><code>", next.(), "</pre></code>\n"]
  end

  def code_line(%Context{} = _context, next, _) do
    [next.(), "\n"]
  end

  def blockquote(%Context{} = _context, next, _) do
    ["<quote>", next.(), "</quote>\n"]
  end

  def a(%Context{} = _context, next, %{"href" => href}) do
    ["<link href=\"#{escape_xml!(href)}\">", next.(), "</link>\n"]
  end

  def definition(%Context{} = _context, next, _) do
    ["<extra>", next.(), "</extra>\n"]
  end

  def text(%Context{} = _context, %{"text" => text} = text_entity) do
    escape_xml!(text) |> wrap_with_marks(text_entity)
  end

  def escape_xml!(text) do
    case HTML.html_escape(text) do
      {:safe, result} -> result
    end
  end

  def wrap_with_marks(text, text_entity) do
    supported_mark_tags = %{
      "em" => "em",
      "strong" => "strong",
      "mark" => "mark",
      "del" => "del",
      "var" => "var",
      "code" => "code",
      "sub" => "sub",
      "sup" => "sup",
    }
    marks = Map.keys(text_entity)
      # only include marks that are set to true
      |> Enum.filter(fn attr_name -> Map.get(text_entity, attr_name) == true end)
      # convert mark to tag name
      |> Enum.map(fn attr_name -> Map.get(supported_mark_tags, attr_name) end)
      # filter out any unsupported marks
      |> Enum.filter(fn mark -> mark != nil end)

    Enum.reverse(marks)
    |> Enum.reduce(
      text,
      fn mark, acc ->
        "<#{mark}>#{acc}</#{mark}>"
      end
    )
  end
end

File: assets/src/data/content/writers/html.ts
import { WriterImpl, Next } from './writer';
import {
  ModelElement,
  Image,
  HeadingSix,
  Paragraph,
  HeadingOne,
  HeadingTwo,
  HeadingThree,
  HeadingFour,
  HeadingFive,
  YouTube,
  Audio,
  Table,
  TableRow,
  TableHeader,
  TableData,
  OrderedList,
  UnorderedList,
  ListItem,
  Math,
  MathLine,
  Code,
  CodeLine,
  Blockquote,
  Hyperlink,
  Webpage,
} from '../model';
import { Text } from 'slate';
import { WriterContext } from './context';

// Important: any changes to this file must be replicated
// in content/html.ex for non-activity rendering.

function escapeHtml(unsafe: string): string {
  return unsafe
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&#039;');
}

export class HtmlParser implements WriterImpl {
  private escapeXml = (text: string) => decodeURI(encodeURI(text));

  private wrapWithMarks(text: string, textEntity: Text): string {
    const supportedMarkTags: { [key: string]: string } = {
      em: 'em',
      strong: 'strong',
      mark: 'mark',
      del: 'del',
      var: 'var',
      code: 'code',
      sub: 'sub',
      sup: 'sup',
    };
    return Object.keys(textEntity)
      .filter((attr) => textEntity[attr] === true)
      .map((attr) => supportedMarkTags[attr])
      .filter((mark) => mark)
      .reduce((acc, mark) => `<${mark}>${acc}</${mark}>`, text);
  }

  private figure(attrs: any, content: string) {
    if (!attrs.caption) {
      return content;
    }

<<<<<<< HEAD
    return `<div class="figure-wrapper ${this.displayClass(attrs)}">
        <figure${
          attrs['full-width'] ? ' class="full-width embed-responsive embed-responsive-16by9"' : ''
        }>
=======
    return `<div class="figure-wrapper">
        <figure class="figure embed-responsive text-center">
>>>>>>> e1e54a7e6c445396b2c7b4dc37e7acf9bac02fa1
          ${content}
          <figcaption class="figure-caption text-center">
            ${this.escapeXml(attrs.caption)}
          </figcaption>
        </figure>
      </div>`;
  }

  p = (context: WriterContext, next: Next, x: Paragraph) => `<p>${next()}</p>\n`;
  h1 = (context: WriterContext, next: Next, x: HeadingOne) => `<h1>${next()}</h1>\n`;
  h2 = (context: WriterContext, next: Next, x: HeadingTwo) => `<h2>${next()}</h2>\n`;
  h3 = (context: WriterContext, next: Next, x: HeadingThree) => `<h3>${next()}</h3>\n`;
  h4 = (context: WriterContext, next: Next, x: HeadingFour) => `<h4>${next()}</h4>\n`;
  h5 = (context: WriterContext, next: Next, x: HeadingFive) => `<h5>${next()}</h5>\n`;
  h6 = (context: WriterContext, next: Next, x: HeadingSix) => `<h6>${next()}</h6>\n`;
  img = (context: WriterContext, next: Next, attrs: Image) => {
    const alt = attrs.alt ? ` alt="${this.escapeXml(attrs.alt)}"` : '';
    const width = attrs.width ? ` width="${this.escapeXml(String(attrs.width))}"` : '';
    const height = attrs.height ? ` height="${this.escapeXml(String(attrs.height))}"` : '';
    return this.figure(
      attrs,
      `<img class="figure-img img-fluid"${alt}${width}${height} src="${this.escapeXml(
        attrs.src,
      )}"/>\n`,
    );
  };
  youtube = (context: any, next: Next, attrs: YouTube) => {
    return this.iframe(context, next, {
      ...attrs,
      src: `https://www.youtube.com/embed/${this.escapeXml(attrs.src)}`,
    });
  };
  iframe = (context: any, next: Next, attrs: Webpage | YouTube) => {
    return this.figure(
      attrs,
      `<div class="embed-responsive embed-responsive-16by9">
          <iframe class="embed-responsive-item" allowfullscreen src="${this.escapeXml(
            attrs.src,
          )}"></iframe>
        </div>`,
    );
  };
  audio = (context: WriterContext, next: Next, attrs: Audio) =>
    this.figure(
      attrs,
      `<audio controls src="${this.escapeXml(
        attrs.src,
      )}">Your browser does not support the <code>audio</code> element.</audio>\n`,
    );
  table = (context: WriterContext, next: Next, attrs: Table) => {
    const caption = attrs.caption ? `<caption>${this.escapeXml(attrs.caption)}</caption>` : '';

    return `<table>${caption}${next()}</table>\n`;
  };
  tr = (context: WriterContext, next: Next, x: TableRow) => `<tr>${next()}</tr>\n`;
  th = (context: WriterContext, next: Next, x: TableHeader) => `<th>${next()}</th>\n`;
  td = (context: WriterContext, next: Next, x: TableData) => `<td>${next()}</td>\n`;
  ol = (context: WriterContext, next: Next, x: OrderedList) => `<ol>${next()}</ol>\n`;
  ul = (context: WriterContext, next: Next, x: UnorderedList) => `<ul>${next()}</ul>\n`;
  li = (context: WriterContext, next: Next, x: ListItem) => `<li>${next()}</li>\n`;
  math = (context: WriterContext, next: Next, x: Math) => `<div>${next()}</div>\n`;
  mathLine = (context: WriterContext, next: Next, x: MathLine) => `${next()}\n`;
  code = (context: WriterContext, next: Next, attrs: Code) =>
    this.figure(
      attrs,
      `<pre><code class="language-${this.escapeXml(attrs.language)}">${next()}</code></pre>\n`,
    );
  codeLine = (context: WriterContext, next: Next, x: CodeLine) => `${next()}\n`;
  blockquote = (context: WriterContext, next: Next, x: Blockquote) =>
    `<blockquote>${next()}</blockquote>\n`;
  a = (context: WriterContext, next: Next, { href }: Hyperlink) => {
    if (href.startsWith('/course/link/')) {
      let internalHref = href;
      if (context.sectionSlug) {
        const revisionSlug = href.replace(/^\/course\/link\//, '');
        internalHref = `/sections/${context.sectionSlug}/page/${revisionSlug}`;
      } else {
        internalHref = '#';
      }

      return `<a class="internal-link" href="${this.escapeXml(internalHref)}">${next()}</a>\n`;
    }

    // tslint:disable-next-line:max-line-length
    return `<a class="external-link" href="${this.escapeXml(
      href,
    )}" target="_blank">${next()}</a>\n`;
  };
  text = (context: WriterContext, textEntity: Text) =>
    this.wrapWithMarks(escapeHtml(textEntity.text), textEntity);

  unsupported = (context: WriterContext, { type }: ModelElement) =>
    '<div class="content invalid">Content element is invalid</div>\n';
}
